code/modules/holodeck/computer.dm

```diff
@@ -216,4 +174,122 @@
 
 /obj/machinery/computer/holodeck/blob_act(obj/structure/blob/B)
 	emergency_shutdown()
-	return ..()
\ No newline at end of file
+	return ..()
+
+/obj/machinery/computer/holodeck/proc/generate_program_list()
+	for(var/typekey in subtypesof(program_type))
+		var/area/holodeck/A = locate(typekey) in GLOB.sortedAreas
+		if(!A || A == offline_program || !A.contents.len)
+			continue
+		var/list/info_this = list()
+		info_this["name"] = A.name
+		info_this["type"] = A.type
+		if(A.restricted)
+			LAZYADD(emag_programs, list(info_this))
+		else
+			LAZYADD(program_cache, list(info_this))
+
+/obj/machinery/computer/holodeck/proc/toggle_power(toggleOn = FALSE)
+	if(active == toggleOn)
+		return
+
+	if(toggleOn)
+		if(last_program && last_program != offline_program)
+			addtimer(CALLBACK(src, .proc/load_program, last_program, TRUE), 25)
+		active = TRUE
+	else
+		last_program = program
+		load_program(offline_program, TRUE)
+		active = FALSE
+
+/obj/machinery/computer/holodeck/proc/emergency_shutdown()
+	last_program = program
+	load_program(offline_program, TRUE)
+	active = FALSE
+
+/obj/machinery/computer/holodeck/proc/floorcheck()
+	for(var/turf/T in linked)
+		if(!T.intact || isspaceturf(T))
+			return FALSE
+	return TRUE
+
+/obj/machinery/computer/holodeck/proc/nerf(active)
+	for(var/obj/item/I in spawned)
+		I.damtype = active ? STAMINA : initial(I.damtype)
+	for(var/e in effects)
+		var/obj/effect/holodeck_effect/HE = e
+		HE.safety(active)
+
+/obj/machinery/computer/holodeck/proc/load_program(area/A, force = FALSE, add_delay = TRUE)
+	if(stat)
```

Comment:
Not entirely sure what this check is, but if you're checking to see if it's working, use is_operational()